retworkx-core/src/traversal/bfs_visit.rs

```diff
@@ -0,0 +1,254 @@
+/// use retworkx_core::petgraph::visit::Control;
+///
+/// use retworkx_core::traversal::{BfsEvent, breadth_first_search};
+///
+/// let gr: Graph<(), ()> = Graph::from_edges(&[
+///     (0, 1), (0, 2), (0, 3),
+///     (1, 3),
+///     (2, 3), (2, 4),
+///     (4, 0), (4, 5),
+/// ]);
+///
+/// // record each predecessor, mapping node → node
+/// let mut predecessor = vec![NodeIndex::end(); gr.node_count()];
+/// let start = n(0);
+/// let goal = n(5);
+/// breadth_first_search(&gr, Some(start), |event| {
+///     if let BfsEvent::TreeEdge(u, v, _) = event {
+///         predecessor[v.index()] = u;
+///         if v == goal {
+///             return Control::Break(v);
+///         }
+///     }
+///     Control::Continue
+/// });
+///
+/// let mut next = goal;
+/// let mut path = vec![next];
+/// while next != start {
+///     let pred = predecessor[next.index()];
+///     path.push(pred);
+///     next = pred;
+/// }
+/// path.reverse();
+/// assert_eq!(&path, &[n(0), n(2), n(4), n(5)]);
+/// ```
+///
+/// # Example returning a `Result`.
+/// ```
+/// use retworkx_core::petgraph::graph::node_index as n;
+/// use retworkx_core::petgraph::prelude::*;
+///
+/// use retworkx_core::traversal::{BfsEvent, breadth_first_search};
+///
+/// let gr: Graph<(), ()> = Graph::from_edges(&[(0, 1), (1, 2), (1, 1), (2, 1)]);
+/// let start = n(0);
+/// let mut non_tree_edges = 0;
+///
+/// #[derive(Debug)]
+/// struct NonTreeEdgeFound {
+///     source: NodeIndex,
+///     target: NodeIndex,
+/// }
+///
+/// // Stop the search, the first time a BackEdge is encountered.
+/// let result = breadth_first_search(&gr, Some(start), |event| {
+///     match event {
+///         BfsEvent::NonTreeEdge(u, v, _) => {
+///             non_tree_edges += 1;
+///             // the implementation of ControlFlow for Result,
+///             // treats this Err value as Continue::Break
+///             Err(NonTreeEdgeFound {source: u, target: v})
+///         }
+///         // In the cases where Ok(()) is returned,
+///         // Result falls back to the implementation of Control on the value ().
+///         // In the case of (), this is to always return Control::Continue.
+///         // continuing the search.
+///         _ => Ok(()),
+///     }
+/// });
+///
+/// assert_eq!(non_tree_edges, 1);
+/// println!("number of non-tree edges encountered: {}", non_tree_edges);
+/// println!("non-tree edge: ({:?})", result.unwrap_err());
+/// ```
+pub fn breadth_first_search<G, I, F, C>(
+    graph: G,
+    starts: I,
+    mut visitor: F,
+) -> C
+where
+    G: IntoEdges + Visitable,
+    I: IntoIterator<Item = G::NodeId>,
+    F: FnMut(BfsEvent<G::NodeId, &G::EdgeWeight>) -> C,
+    C: ControlFlow,
+{
+    let discovered = &mut graph.visit_map();
+    let finished = &mut graph.visit_map();
+
+    for start in starts {
+        try_control!(
+            bfs_visitor(graph, start, &mut visitor, discovered, finished),
+            unreachable!()
```

Comment:
I think having a comment here would be nice to explain why this branch is unreachable. While it's pretty obvious from the context for these kind of things I find it better to be exlicit

Reply:
Done in aa27296